u2f-ref-code/java/src/com/google/u2f/tools/X509Util.java

```
@@ -0,0 +1,151 @@
+      return false;
+    }
+
+    // Walk through intermediates up the chain
+    try {
+      for (int i = 0; i < certChain.length - 1; i++) {
+        certChain[i].verify(certChain[i + 1].getPublicKey());
+      }
+    } catch (SignatureException | NoSuchProviderException | InvalidKeyException
+        | NoSuchAlgorithmException | CertificateException e) {
+      Log.log(Level.SEVERE, "Cannot validate cert chain is correctly signed by intermediaries", e);
+      return false;
+    }
+
+    // Now attempt to verify up to one of the roots
+    boolean validated = true;
+    for (int i = 0; i < caCerts.length; i++) {
+      try {
+        certChain[certChain.length - 1].verify(caCerts[i].getPublicKey());
+      } catch (SignatureException | NoSuchProviderException | InvalidKeyException
+          | NoSuchAlgorithmException | CertificateException e) {
+        Log.log(Level.WARNING, "Cert chain validation failed to match a root.", e);
+
+        // If it's the last cert, that means we didn't find a matching root ca
+        if (i == caCerts.length - 1) {
+          validated = false;
+          Log.log(Level.SEVERE, "Cannot validate cert chain is correctly signed by any known root");
+        }
+      }
+    }
+
+    return validated;
+  }
+
+  /**
+   * Encodes an array of certificates into a DER-encoded byte array
+   * @param certsArray to encode
+   * @return the encoded certs
+   */
+  public static byte[] encodeCertArray(X509Certificate[] certsArray) {
+    if (certsArray == null) {
+      return null;
+    }
+
+    if (certsArray.length == 0) {
+      return new byte[0];
+    }
+
+    try {
+      ByteArrayOutputStream encodedCerts = new ByteArrayOutputStream();
+      for (X509Certificate cert : certsArray) {
+        encodedCerts.write(cert.getEncoded());
+      }
+      return encodedCerts.toByteArray();
+    } catch (CertificateEncodingException | IOException e) {
+      throw new RuntimeException();
```

Please propagate the exception in this constructor.